extensions/protobuf/src/main/java/io/deephaven/protobuf/ProtobufDescriptorParserImpl.java

```diff
@@ -0,0 +1,541 @@
+        }
+
+        private ProtobufFunctions functions() {
+            final ProtobufFunctions wellKnown = wellKnown().orElse(null);
+            if (wellKnown != null) {
+                return wellKnown;
+            }
+            final Builder builder = ProtobufFunctions.builder();
+            for (FieldContext fc : fcs()) {
+                builder.addAllFunctions(fc.functions().functions());
+            }
+            return builder.build();
+        }
+
+        private Optional<ProtobufFunctions> wellKnown() {
+            // todo: eventually support cases that are >1 field
+            final SingleValuedMessageParser svmp = byFullName.get(descriptor.getFullName());
+            if (svmp == null) {
+                return Optional.empty();
+            }
+            if (!options.parseAsWellKnown().test(fieldPath)) {
+                return Optional.empty();
+            }
+            return Optional.of(ProtobufFunctions.unnamed(svmp.messageParser(descriptor, options)));
+        }
+
+        private List<FieldContext> fcs() {
+            return descriptor.getFields().stream().map(this::fc).collect(Collectors.toList());
+        }
+
+        private FieldContext fc(FieldDescriptor fd) {
+            return new FieldContext(this, fd);
+        }
+    }
+
+    private class FieldContext {
+        private final DescriptorContext parent;
+        private final FieldDescriptor fd;
+        private final FieldPath fieldPath;
+
+        public FieldContext(DescriptorContext parent, FieldDescriptor fd) {
+            this.parent = Objects.requireNonNull(parent);
+            this.fd = Objects.requireNonNull(fd);
+            this.fieldPath = parent.fieldPath.append(fd);
+        }
+
+        private ProtobufFunctions functions() {
+            return functions(false);
+        }
+
+        private ProtobufFunctions functions(boolean forceInclude) {
+            if (!forceInclude && !options.include().test(fieldPath)) {
+                return ProtobufFunctions.empty();
+            }
+            final ProtobufFunctions wellKnown = wellKnown().orElse(null);
+            if (wellKnown != null && options.parseAsWellKnown().test(fieldPath)) {
+                return wellKnown;
+            }
+            if (fd.isMapField() && options.parseAsMap().test(fieldPath)) {
+                return new MapFieldObject().functions();
+            }
+            if (fd.isRepeated()) {
+                return new RepeatedFieldObject().functions();
+            }
+            return new FieldObject().functions();
+        }
+
+        private Optional<ProtobufFunctions> wellKnown() {
+            // todo: eventually have support for parsing specific fields in specific ways
+            return Optional.empty();
+        }
+
+        private ProtobufFunctions namedField(TypedFunction<Message> tf) {
+            // todo: can we re-work this, so we can use fieldPath here instead and not use prefix later?
```

Comment:
do you want to track this with an issue?

Reply:
Implementation detail, not really deserving of issue. Removed this comment.